File: welvet/python/examples/04_mha_forward.py
#!/usr/bin/env python3
"""Example 4: Multi-head attention with [batch, seq, d_model] layout."""

import math

from welvet import Network


def main() -> None:
    batch, seq, d_model, heads = 2, 4, 32, 4
    in_shape = [batch, seq, d_model]

    net = Network({
        "id": "example-mha",
        "depth": 1,
        "rows": 1,
        "cols": 1,
        "layers_per_cell": 1,
        "layers": [{
            "z": 0, "y": 0, "x": 0, "l": 0,
            "type": "mha",
            "dtype": "float32",
            "d_model": d_model,
            "num_heads": heads,
            "seq_length": seq,
            "activation": "relu",
        }],
    })

    n = batch * seq * d_model
    inp = [0.15 * math.sin(i * 0.09) for i in range(n)]
    out = net.forward_polymorphic(inp, in_shape)

    assert len(out) == n, f"expected {n} outputs, got {len(out)}"
    assert all(math.isfinite(x) for x in out[:8])

    net.free()
    print("04_mha_forward OK — shape=%s out[0]=%.4f" % (in_shape, out[0]))


if __name__ == "__main__":
    main()

File: welvet/dart/tool/examples/04_mha_forward.dart
// Example 4: MHA forward with [batch, seq, d_model].
//
//   cd examples && dart run 04_mha_forward.dart

import 'dart:convert';
import 'dart:math' as math;

import 'package:welvet/loom_ffi.dart';

void main() {
  if (!loomAvailable) {
    print('Welvet not loaded: $loomLibLastError');
    return;
  }

  const batch = 4;
  const seq = 8;
  const dModel = 64;
  const heads = 4;

  final netJson = jsonEncode({
    'id': 'example-mha',
    'depth': 1,
    'rows': 1,
    'cols': 1,
    'layers_per_cell': 1,
    'layers': [
      {
        'z': 0,
        'y': 0,
        'x': 0,
        'l': 0,
        'type': 'MHA',
        'dtype': 'FLOAT32',
        'd_model': dModel,
        'num_heads': heads,
        'seq_length': seq,
        'activation': 'RELU',
      },
    ],
  });

  final handle = loomLib.createNetwork(netJson);
  final inp = List<double>.generate(
    batch * seq * dModel,
    (i) => 0.2 * math.sin(i * 0.11),
  );
  final inShape = [batch, seq, dModel];

  loomLib.configureTrainingMode(handle, 2);
  final out = loomParseFloatArray(
    loomLib.forwardPolymorphic(handle, inp, inShape),
  );
  if (out == null || out.isEmpty) throw StateError('MHA forward failed');

  loomLib.freeNetwork(handle);
  print('04_mha_forward OK — out_len=${out.length} out[0]=${out[0].toStringAsFixed(4)}');
}
